flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java

```diff
@@ -547,4 +549,30 @@ public boolean accept(File file, String s) {
 			return true;
 		}
 	}
+
+	private static class TestRocksDBStateBackend extends RocksDBStateBackend {
+
+		public TestRocksDBStateBackend(AbstractStateBackend checkpointStreamBackend, boolean enableIncrementalCheckpointing) {
+			super(checkpointStreamBackend, enableIncrementalCheckpointing);
+		}
+
+		@Override
+		public <K> AbstractKeyedStateBackend<K> createKeyedStateBackend(
+			Environment env,
+			JobID jobID,
+			String operatorIdentifier,
+			TypeSerializer<K> keySerializer,
+			int numberOfKeyGroups,
+			KeyGroupRange keyGroupRange,
+			TaskKvStateRegistry kvStateRegistry) throws IOException {
+
+			AbstractKeyedStateBackend<K> keyedStateBackend = super.createKeyedStateBackend(
+				env, jobID, operatorIdentifier, keySerializer, numberOfKeyGroups, keyGroupRange, kvStateRegistry);
+
+			// We ignore the range deletions on production, but when we are running the tests we shouldn't ignore it.
```

Comment:
Can you briefly explain why this whole `ReadOptions` change is required and what this comment about ignoring range deleted is related to? This seems to introduce some implicit complexities, so I just want to double check if this is really required.

Reply:
I introduce this because we may called the `deleteRange()` when rescaling from incremental checkpoint. According to the RocksDB's comment on its CPP file, in order to get rid of the downside of the read performance, we should set `readOptions.setIgnoreRangeDeletions(true);`. 

- On production, that is fine because we won't query any record that belong to the key-group we have delete.

- But when running tests, we may need to verify that after restoring from checkpoint, we didn't take any external key-group that isn't belong to the target key group into the backend(e.g. `StateBackendTestBase#testKeyGroupSnapshotRestore().`). The reason that we need to `readOptions.setIgnoreRangeDeletions(false);` in this case might be explain as below:

```java
db.deleteRange(range1);
readOptions.setIgnoreRangeDeletions(true);

db.get(readOptions, key in range1); // this may not be null, because we have ignore the range deletions.

readOptions.setIgnoreRangeDeletions(false);

db.get(readOptions, key in range1); // this will be null
```

Reply:
As I see, this is only happening in the case where there is only one handle and we are only interested in a subset of the key-groups. Unfortunately, that should be the common case of scaling out. I am wondering if we should not prefer to apply normal deletes over range delete, because what will happen if we take again a snapshot from a database that was using range deletes? Will the keys all be gone in cases of full and incremental snapshots? If the performance of normal deletes is not terrible, that might be cleaner for as long as range deletes are not working properly or have potential negative side-effects. What is your opinion about this?

Reply:
Sorry...I think I may didn't understand "I am wondering if we should not prefer to apply normal deletes over range delete" properly, is that mean "I am wondering if we should prefer to apply normal deletes over range delete". As far as I know the keys all be gone only when compaction occur, for `deleteRange()` it only write a special record in db, looks like `Deleted Range(beginKey, endKey]`, it won't remove any records from the db indeed.

And yes, concerning to the negative side-effects of the `deleteRange()` I also still have the same concerns, even thought rescaling from checkpoint is still an experimental feature. I think a more safer way to improve the performance of recover from incremental checkpoint is we don't clip it, and only choose the instance to be the initial db when its key-group range is a subset of the target key-group range. What do you think?

Reply:
Or maybe, the `chooseTheBestStateHandleToInit` could also return a non-null instance if it's found the there are only a few invalid key-group that doesn't belong to the target key-group, and comparing to iterate the valid records and put them, it more cheap to just delete the invalid records, e.g. the target key-group range is [0, 10] and there is one handle's key-group is [2, 11], this case we may only need to delete(use the normal delete) the key-group 11 from this handle.

Reply:
Yes, I think that makes sense. My first question was if you think that using normal deletes would be prohibitive expensive. I think there is some threshold where restore+single deletes is cheaper than no restore+filtered inserts. So the question is, for up to how much of the key group fraction which strategy is the best. Do you think the single deletes are just suitable to remove only like 1-2 key-groups or up to which point will this be better than starting with an empty database?

Reply:
Oh so poor of my english...I think maybe base on fraction is a better choose, e.g. if `"number of the invalid key-group of handle" / "number of all key-group of handle" <= 1/4`, we prefer to the "restore+single deletes".

Reply:
That sounds good 👍 This also means we do not need any tricks with the read options :)

Reply:
Yes, I will change it according to the approach we discussion above.

Reply:
On a different note, this became more complex now, I wonder if we should also add a test for incremental rescale. I think that could be done at a level of using the `KeyedOneInputStreamOperatorTestHarness` for different ranges, choosing incremental RocksDB, trigger two checkpoints, scale in and out into a new harness with different key group range. Ideally, this test could cover all the different corner cases (about the fraction)

Reply:
I also think it needs a test now, I will add it.